src/game/puppyprint.c

```diff
@@ -770,105 +961,218 @@ const Gfx dl_small_text_begin[] = {
-            textPos[0] += (spaceX + 1);
+            if (i >= strLen)
+                break;
+
+            get_char_from_byte(str[i], &textX, &spaceX, &offsetY, font);
+            textPos[0] += (spaceX + 1) * textSizeTotal;
             wideX[lines] = MAX(textPos[0], wideX[lines]);
         }
 
-        textPos[0] = -(wideX[0] / 2);
-    } else if (align == PRINT_TEXT_ALIGN_RIGHT) {
-        for (i = 0; i < (signed)strlen(str); i++) {
-            if (str[i] == '#') {
-                i++;
-                textPos[0] = 0;
-                lines++;
-            } else {
-                textPos[0] += (spaceX + 1);
-            }
-
-            if (str[i] == '<') {
-                i += text_iterate_command(str, i, FALSE);
-            }
-
-            get_char_from_byte(str[i], &textX, &textY, &spaceX, &offsetY, font);
-
-            wideX[lines] = MAX(textPos[0], wideX[lines]);
+        if (align == PRINT_TEXT_ALIGN_CENTRE) {
+            textPos[0] = -(wideX[0] / 2);
+        } else {
+            textPos[0] = -(wideX[0]);
         }
-        textPos[0] = -wideX[0];
     }
 
+    //Reset text size properties.
+    textSizeTemp = 1.0f;
+    set_text_size_params();
+    topLineHeight = 12.0f * textSizeTotal;
     lines = 0;
-    gDPLoadTextureBlock_4b(gDisplayListHead++, (*fontTex)[font], G_IM_FMT_I, 128, 60, (G_TX_NOMIRROR | G_TX_CLAMP), (G_TX_NOMIRROR | G_TX_CLAMP), 0, 0, 0, G_TX_NOLOD, G_TX_NOLOD);
-    for (i = 0; i < tx; i++) {
-        if (str[i] == '#') {
-            i++;
+    
+    shakeTablePos = gGlobalTimer % sizeof(sTextShakeTable);
+    gDPLoadTextureBlock_4b(gDisplayListHead++, (*fontTex)[font], G_IM_FMT_I, 672, 12, (G_TX_NOMIRROR | G_TX_CLAMP), (G_TX_NOMIRROR | G_TX_CLAMP), 0, 0, 0, G_TX_NOLOD, G_TX_NOLOD);
+    
+    for (s32 i = 0, j = 0; i < textLength; i++, j++) {
+        if (str[i] == '#' || str[i] == 0x0A) {
             lines++;
             if (align == PRINT_TEXT_ALIGN_RIGHT) {
-                textPos[0] = -(wideX[lines]    );
+                textPos[0] = -(wideX[lines]);
             } else {
                 textPos[0] = -(wideX[lines] / 2);
             }
-            textPos[1] += 12;
+            textPos[1] += topLineHeight;
+            topLineHeight = 12.0f * textSizeTotal;
+            continue;
         }
 
-        if (str[i] == '<') {
-            i += text_iterate_command(str, i, TRUE);
+        while (i < textLength && str[i] == '<') {
+            commandOffset = text_iterate_command(str, i, TRUE);
+            if (commandOffset == 0)
+                break;
+
+            i += commandOffset;
         }
 
+        if (i >= textLength)
+            break;
+
         if (shakeToggle) {
-            shakePos[0] = (-1 + (random_u16() & 0x1));
-            shakePos[1] = (-1 + (random_u16() & 0x1));
+            shakePos[0] = sTextShakeTable[shakeTablePos++] * textSizeTotal;
+            if (shakeTablePos == sizeof(sTextShakeTable))
```

Comment:
Also since it's unlikely the shake table will be changed to some random size by the user, does it make more sense to turn this into a modulo instead or no? Assuming the compiler isn't smart enough to optimize this into a `&= 0x1F` in this case.